rust/src/api.rs

```diff
@@ -139,13 +115,121 @@ pub fn decode_invoice(invoice: String) -> Result<LightningInvoice> {
+        let lightning_history = wallet::get_lightning_history().await?;
+        Ok(WalletInfo {
+            balance,
+            bitcoin_history,
+            lightning_history,
+        })
+    }
+}
+
 #[tokio::main(flavor = "current_thread")]
-pub async fn run_ldk() -> Result<()> {
+pub async fn refresh_wallet_info() -> Result<WalletInfo> {
+    WalletInfo::build_wallet_info().await
+}
+
+#[tokio::main(flavor = "current_thread")]
+pub async fn run(stream: StreamSink<Event>, app_dir: String) -> Result<()> {
+    let network = config::network();
+    anyhow::ensure!(!app_dir.is_empty(), "app_dir must not be empty");
+    stream.add(Event::Init(format!("Initialising {network} wallet")));
+    wallet::init_wallet(Path::new(app_dir.as_str()))?;
+
+    stream.add(Event::Init("Initialising database".to_string()));
+    db::init_db(
+        &Path::new(app_dir.as_str())
+            .join(network.to_string())
+            .join("taker.sqlite"),
+    )
+    .await?;
+
     tracing::debug!("Starting ldk node");
+    stream.add(Event::Init("Starting full lightning node".to_string()));
     match wallet::run_ldk().await {
         Ok(_background_processor) => {
+            stream.add(Event::Init("Fetching an offer!".to_string()));
+            stream.add(Event::Offer(offer::get_offer().await));
+            stream.add(Event::Init("Fetching your balance".to_string()));
+            stream.add(Event::WalletInfo(
+                WalletInfo::build_wallet_info()
+                    .await
+                    .unwrap_or_else(|_| WalletInfo::new()),
+            ));
+
+            stream.add(Event::Ready);
             // keep connection with maker alive!
-            connection::keep_alive().await?;
+            let mut connected = false;
+
+            let mut wallet_sync_counter = 60;
+            let mut wallet_info_counter = 10;
+            loop {
+                if !connected || !is_first_channel_usable() {
+                    connected = connect().await?;
+                }
+
+                // every 5 seconds
+                stream.add(Event::Offer(offer::get_offer().await));
+
+                // every 60 seconds
+                if wallet_sync_counter == 60 {
+                    wallet::sync().unwrap_or_else(|e| tracing::error!(?e, "Failed to sync wallet"));
+                    wallet_sync_counter = 0;
+                }
+
+                // every 10 seconds
+                if wallet_info_counter == 10 {
+                    match WalletInfo::build_wallet_info().await {
+                        Ok(wallet_info) => {
+                            stream.add(Event::WalletInfo(wallet_info));
+                        }
+                        Err(e) => {
+                            tracing::error!(?e, "Failed to build wallet info");
+                        }
+                    }
+                    wallet_info_counter = 0;
+                }
+
+                wallet_sync_counter += 5;
+                wallet_info_counter += 5;
```

Comment:
🤔 Depending on what actions were performed in the current loop, a lot more than 5 seconds may have passed.

In general, it's unfortunate that we cannot have separate loops for each task, as the logic would be simpler and less error-prone.

Reply:
that's very true. For the get offer and get balance, this delay is actually not noticeable. However, the wallet sync could become problematic (as it seems to take more time the more transactions are performed). We could time the sync and remove it from the timeout? (very hacky) or invest into a dedicated thread for syncing (note, i am still not completely sure how many threads are available to use - the limitation might just apply to bridge calls and not to threads spawned from within a rust function - i have to check)

Reply:
just had a chat with @klochowicz about the topic. The 4 threads are only limited to the bridge calls and pose no limits on the rust code in general. So we can spawn dedicated threads for syncing.

Reply:
https://github.com/itchysats/10101/pull/550/commits/cc7dccd2fdcfb3216744ddcfcd10ce3dbdaffa36

Reply:
I would like to note that the commit mentioned above spawns dedicated tasks on the same single-threaded runtime - not separate threads.

Reply:
I guess that's due to this peace 

```
#[tokio::main(flavor = "current_thread")]
```

Can we use another flavor?

Reply:
I've never researched how a different flavour would work with `frb`.